accepted/0000-dependency-selector-syntax.md

```diff
@@ -0,0 +1,421 @@
+  
+// any workspace dep
+.workspace
+
+// all workspaces that depend on another workspace
+.workspace > .workspace
+    
+// all workspaces that have peer deps
+.workspace:has(*.peer)
+
+// any dep named "lodash"
+// equivalent to [name="lodash"]
+#lodash 
+  
+// any deps named "lodash" & within semver range ^"1.2.3"
+#lodash@^1.2.3
+// equivalent to...
+[name="lodash"]:semver(^1.2.3)
+
+// get the hoisted node for a given semver range
+#lodash@^1.2.3:not(:deduped)
+
+// querying deps with a specific version
+#lodash@2.1.5
+// equivalent to...
+[name="lodash"][version="2.1.5"]
+  
+// all deps living alongside vulnerable deps
+*:vulnerable ~ *:not(:vulnerable)
+
+// has any deps
+*:has(*)
+
+// has any vulnerable deps
+*:has(*:vulnerable)
+
+// deps with no other deps (ie. "leaf" nodes)
+*:empty
+
+// all vulnerable deps that aren't dev deps & that aren't vulnerable to CWE-1333
+*:vulnerable:not(.dev:cwe(1333))
+  
+// manually querying git dependencies
+*[repository^="github:"],
+*[repository^="git:"],
+*[repository^="https://github.com"],
+*[repository^="http://github.com"],
+*[repository^="https://github.com"],
+*[repository^="+git:..."]
+  
+// querying for all git dependencies
+*:type(git)
+
+// find all references to "install" scripts
+*[scripts=install], 
+*[scripts=postinstall],
+*[scripts=preinstall]
+
+// get production dependencies that aren't also dev deps
+.prod:not(.dev)
+    
+// get dependencies with specific licenses
+*[license="MIT"], *[license="ISC"]
+
+// find all packages that have @ruyadorno as a contributor
+*:contributors([email=ruyadorno@github.com])
+```
+
+## Next Steps
+
+### Command Mapping to `query-output`
+
+Previous commands with similar behaivours will now be able to utilize `Aborist` `Node.querySelectorAll()` under-the-hood & will fast-follow the `npm query` implementation.
+
+#### `npm list`
+
+```bash
+npm list # equivalent to...
+npm query ":root > *"
+
+npm list --all # equivalent to...
+npm query "*" --query-output list
+
+npm list <pkg> # equivalent to...
+npm query "#<pkg>" --query-output list
+```
+
+#### `npm explain`
+
+```bash
+npm explain <pkg> # equivalent to...
+npm query "#<pkg>" --query-output explain
+```
+
+#### `npm outdated`
+
+```bash
+npm outdated # equivalent to...
+npm query ":root > *:outdated" --query-output outdated
+
+npm outdated --all # equivalent to...
+npm query "*:outdated" --query-output outdated
+```
+
+#### `npm fund`
+
+```bash
+npm fund # equivalent to...
+npm query ":root > *[funding]" --query-output funding
+
+npm fund --all # equivalent to...
+npm query "*[funding]" --query-output funding
+
+npm fund <pkg> # equivalent to...
+npm query "#<pkg>" --query-output funding
+```
+
+#### `npm audit`
+
+```bash
+npm audit # equivalent to...
+npm query ":root > *:vulnerable" --query-output audit
+
+npm audit --all # equivalent to...
+npm query "*:vulnerable" --query-output audit
+
+npm audit <pkg> # equivalent to...
+npm query "#<pkg>" --query-output audit
+```
```

Comment:
AFAIK this would also allow for easy implementation of https://github.com/npm/rfcs/pull/182

Reply:
100%